alchemiscale_fah/compute/service.py

```diff
@@ -4,6 +4,621 @@
+from gufe.protocols.protocoldag import (
+    ProtocolDAG,
+    ProtocolDAGResult,
+    _pu_to_pur,
+    Context,
+)
+from gufe.protocols.protocolunit import ProtocolUnit, ProtocolUnitResult
+
+from alchemiscale.models import Scope, ScopedKey
+from alchemiscale.storage.models import Task, TaskHub, ComputeServiceID
+from alchemiscale.compute.client import AlchemiscaleComputeClient
+from alchemiscale.compute.service import (
+    SynchronousComputeService,
+    InterruptableSleep,
+    SleepInterrupted,
+)
+from alchemiscale.keyedchain import KeyedChain
+
+from .models import FahProject
+from .settings import FahAsynchronousComputeServiceSettings
+from .client import FahAdaptiveSamplingClient
+from .index import FahComputeServiceIndex
+from ..protocols.protocolunit import FahSimulationUnit, FahContext
+
+# NOTE: if we don't do this, it appears that our RDKit mols won't keep their
+# properties on being pickled with a ProcessPoolExecutor, despite our efforts to avoid this;
+# this is important for retaining partial charges
+from rdkit import Chem
+
+Chem.SetDefaultPickleProperties(Chem.PropertyPickleOptions.AllProps)
+
+
+class FahAsynchronousComputeService(SynchronousComputeService):
+    """An asynchronous compute service for utilizing a Folding@Home work server."""
+
+    def __init__(self, settings: FahAsynchronousComputeServiceSettings):
+        """Create a `FAHSynchronousComputeService` instance."""
+        self.settings = settings
+
+        self.api_url = self.settings.api_url
+        self.name = self.settings.name
+        self.sleep_interval = self.settings.sleep_interval
+        self.heartbeat_interval = self.settings.heartbeat_interval
+        self.claim_limit = self.settings.claim_limit
+
+        self.client = AlchemiscaleComputeClient(
+            self.settings.api_url,
+            self.settings.identifier,
+            self.settings.key,
+            max_retries=self.settings.client_max_retries,
+            retry_base_seconds=self.settings.client_retry_base_seconds,
+            retry_max_seconds=self.settings.client_retry_max_seconds,
+            verify=self.settings.client_verify,
+        )
+
+        self.fah_project_ids = self.settings.fah_project_ids
+
+        self.index_dir = Path(self.settings.index_dir).absolute()
+        self.obj_store = Path(self.settings.obj_store).absolute()
+
+        self.fah_client = FahAdaptiveSamplingClient(
+            as_url=self.settings.fah_as_url,
+            ws_url=self.settings.fah_ws_url,
+            certificate_file=self.settings.fah_certificate_file,
+            key_file=self.settings.fah_key_file,
+            csr_file=self.settings.fah_csr_file,
+            verify=self.settings.fah_client_verify,
+        )
+
+        # get PROJECT data from metadata files in each PROJECT
+        self.fah_projects = [
+            FahProject.parse_raw(
+                self.fah_client.get_project_file_to_bytes(
+                    p, "alchemiscale-project.txt"
```

Comment:
Documentation / format for this file?

Reply:
Must always be in root relative to running process also? Is this checked and caught properly? (Can't remember)

Reply:
This isn't a local file; it's being pulled from the WS API for each PROJECT the service is configured to use.

Reply:
These files correspond to the `FahProject` model.